### PR TITLE
Use prettier-java for preformatting before writing disk

### PR DESCRIPTION
**Pull request by pascalgrimaud**
1st step for https://github.com/jhipster/generator-jhipster/issues/10641

@murdos : thanks for pointing me in my other PRs. After our lunch yesterday, I gave a try and you're right: it was easier than expected !

I don't think it's a breaking change here as I didn't introduce the hook with Java, only for our generated code. So we won't change the custom code of our users.

What do you think ?
cc @hdurix @murdos @clementdessoude @Shaolans @Hawkurane @jdubois 
cc @jhipster/developers too

_____

-   Please make sure the below checklist is followed for Pull Requests.

-   [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [ ] Tests are added where necessary
-   [ ] Documentation is added/updated where necessary
-   [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by clementdessoude** [comment]
LGTM

**Comment by murdos** [comment]
Before merging this I think we should solve the issue of the formatting of toString() methods, related to the discussion around https://github.com/jhipster/generator-jhipster/pull/11278

**Comment by pascalgrimaud** [comment]
@murdos : yes, that's why I put 'on hold' label

**Comment by PierreBesson** [comment]
It's not a breaking change but I think it will introduce a huge number of minor formatting changes for people who upgrade so that's why I think it's better if we keep this for v7. However this can be the first merged PR of the v7 branch !

**Comment by murdos** [comment]
> @murdos : yes, that's why I put 'on hold' label

Oh, sorry, I had not seen the label.

**Comment by mshima** [comment]
I am in favor of merging this with a cli flag.

**Comment by pascalgrimaud** [comment]
@mshima : do you suggest a new flag `--prettier` ? 
When using `jhipster --prettier`, it will format all generated java file. Otherwise, it will generate java like it is in template

**Comment by jdubois** [comment]
The flag is a great idea, but I would do the opposite: have Prettier by default, and a flag to disable it in case of trouble.

**Comment by deepu105** [comment]
LGTM. But I'm also in favour of doing it in v7 rather than adding more flags

**Comment by jdubois** [comment]
I don't want to wait... Could we do a test on the sample project, in a specific branch, to see what changes? If that's only formatting it shouldn't too annoying, or so I hope!

**Comment by pascalgrimaud** [comment]
@jdubois : for my current customer, we use prettier-java (with pre commit hook) for 2 months now and it works very well. I encounter only 2 minors issues:
- the import is different than IntelliJ but it's so minor and now, we don't care about it
- the toString is not really readeable, that's why I suggested to add `//prettier-ignore` to this method but we didn't manage to agree - see https://github.com/jhipster/generator-jhipster/pull/11278

I was not alone, @hdurix and @murdos did the same.

@deepu105 : the v7 won't be released before at least 6 months, maybe more... as our roadmap is not clear enough (see https://github.com/jhipster/generator-jhipster/issues/10958). We have too much items here and no work in progress.

About this PR, we didn't introduce any break change, that's why I'm suggesting this.
The break change would be to introduce the pre commit hook which will format all existing / custom code of our users, and I didn't do it here. We'll add this pre commit hook only for v7.

It's similar to **manually** format this kind of things:
- https://github.com/jhipster/generator-jhipster/blob/master/generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs#L848
- https://github.com/jhipster/generator-jhipster/blob/master/generators/entity-server/templates/src/main/java/package/service/mapper/BaseEntityMapper.java.ejs#L36-L38

**Comment by mshima** [comment]
> @mshima : do you suggest a new flag `--prettier` ?

I was thinking about `—prettier-java`. Only prettier can be confusing.
But any flag would work for me. `—experimental` too.

> When using `jhipster --prettier`, it will format all generated java file. Otherwise, it will generate java like it is in template

Exactly.

**Comment by pascalgrimaud** [comment]
I added `--skip-prettier-java` instead, so we'll have it by default

**Comment by deepu105** [comment]
Ok lets do it

**Comment by pascalgrimaud** [comment]
Before merging this, we need to discuss about the `toString` formatting.
We need to make a choice:

1) use the new formatting provided by `prettier-java`

```java
    @Override
    public String toString() {
        return (
            "User{" +
            "login='" +
            login +
            '\'' +
            ", firstName='" +
            firstName +
            '\'' +
            ", lastName='" +
            lastName +
            '\'' +
            ", email='" +
            email +
            '\'' +
            ", imageUrl='" +
            imageUrl +
            '\'' +
            ", activated='" +
            activated +
            '\'' +
            ", langKey='" +
            langKey +
            '\'' +
            ", activationKey='" +
            activationKey +
            '\'' +
            "}"
        );
    }
```

2) add `// prettier-ignore` just before the toString method, so it will be similar to what IntelliJ will generate.

```java
    // prettier-ignore
    @Override
    public String toString() {
        return "User{" +
            "login='" + login + '\'' +
            ", firstName='" + firstName + '\'' +
            ", lastName='" + lastName + '\'' +
            ", email='" + email + '\'' +
            ", imageUrl='" + imageUrl + '\'' +
            ", activated='" + activated + '\'' +
            ", langKey='" + langKey + '\'' +
            ", activationKey='" + activationKey + '\'' +
            "}";
    }
```

3) use apache-commons `ToStringBuilder`

```java
    @Override
    public String toString() {
        return new ToStringBuilder(this)
            .append("firstName", firstName)
            .append("lastName", lastName)
            .append("email", email)
            .toString();
    }
```

File: generators/common/templates/.prettierrc.ejs [hunk 1]
```diff
@@ -10,3 +10,9 @@ arrowParens: avoid
 
 # jsx and tsx rules:
 jsxBracketSameLine: false
+
+# java rules:
```
Comment (murdos) [inline generators/common/templates/.prettierrc.ejs]: This should be added conditionally: it's not useful if skipServer is true

File: package.json [hunk 2]
```diff
@@ -71,6 +71,7 @@
         "parse-gitignore": "1.0.1",
         "pluralize": "7.0.0",
         "prettier": "1.19.1",
+        "prettier-plugin-java": "0.7.0",
```
Comment (DanielFran) [inline package.json]: Update version to 0.7.1?

Comment (pascalgrimaud) [reply package.json]: thanks, done

**Comment by pascalgrimaud** [comment]
@jdubois : I'll change my PR to have `—prettier-java` instead. We won't have prettier-java by default as some of us are afraid of breaking change.

Then, after my change, this PR will be ready to be merged.

**Comment by jdubois** [comment]
I agree, but if it's not by default, people won't use it... Let's put it that way for a couple of months, and we need to raise awareness on this.
Some blog posts, documentation, conference talks should be good -> we need to team up on that

**Comment by pascalgrimaud** [comment]
@murdos : is this last version ok for you ?

**Comment by pascalgrimaud** [comment]
Thanks ! I'll update the documentation with this new flag `--prettier-java`

**Pull request merged**